### PR TITLE
Refactor/improved-architecture

### PR DESCRIPTION
**Pull request by raulmarindev**
# OpenModels PR Documentation

## Summary
This PR significantly refactors and expands the OpenModels library, transforming it from a basic JSON serialization tool for scikit-learn models into a flexible, extensible framework for model serialization. The changes introduce a plugin-based architecture, support for multiple serialization formats, and improved type hinting and error handling.

### Core Architecture
- Introduced `SerializationManager` class as the main interface for serialization operations
- Implemented `FormatRegistry` for managing different serialization formats
- Created `ModelSerializer` and `FormatConverter` protocols for extensibility

### Serialization Formats
- Added support for JSON and pickle formats
- Implemented `JSONConverter` and `PickleConverter` classes

### Scikit-learn Support
- Refactored scikit-learn serialization into `SklearnSerializer` class
- Updated supported estimators list and serialization logic

### Error Handling
- Introduced custom exception classes for more specific error handling

### Testing
- Updated test helpers to work with the new architecture
- Improved type hinting in test functions

### Documentation
- Extensively updated README.md with new features, installation instructions, and usage examples
- Added docstrings to all new classes and functions

## Reasoning
1. Extensibility: The new architecture allows easy addition of new serialization formats and model types.
2. Flexibility: Users can now choose between different serialization formats (e.g., JSON for readability, pickle for performance).
3. Type Safety: Improved type hinting enhances code reliability and IDE support.
4. Error Handling: Custom exceptions provide more informative error messages.
5. Maintainability: The modular structure makes the codebase easier to maintain and extend.

## Testing
- Existing tests should be updated to use the new `SerializationManager` interface.
- New tests should be added for:
  - `SerializationManager` class
  - `FormatRegistry` functionality
  - JSON and pickle converters
  - Custom exception handling

## Potential Impact
- Breaking Changes: The API has changed significantly, requiring users to update their code.
- Performance: The new architecture may have a slight overhead compared to the previous direct JSON serialization.
- Compatibility: Ensure backwards compatibility with previously serialized models.

## Additional Notes
- Consider adding a migration guide for users updating from the previous version.
- The README now mentions support for any serialization format, but only JSON and pickle are implemented. Consider clarifying this or adding a roadmap for future format support.

### CODE REVIEW

**Review by Gnpd — APPROVED** [review]

**Review by paucablop — COMMENTED** [review]
Hi @raulmarindev! :v: 

Fantastic job, many thanks for your efforts :star_struck: :star_struck: ! I really enjoy how you have structured the project. It is very clear. I like that the way you have modularized the code, it will become easier to extend the code, both to more models and to more formats :rocket:. I only had a couple of suggestions in the ```sklearn_serializer.py``` :smile:

File: openmodels/serializers/sklearn_serializer.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        ]
+
+        return {
+            "attributes": dict(
+                zip(filtered_attribute_keys, serializable_attribute_values)
+            ),
+            "attribute_types": [str(attr_type) for attr_type in attribute_types],
+            "estimator_class": model.__class__.__name__,
+            "params": model.get_params(),
+            "producer_name": "sklearn",
```
[the remark this answers is not in the document]
Comment (raulmarindev) [reply openmodels/serializers/sklearn_serializer.py:226]: This is a great idea, thanks! I'll implement it.

File: openmodels/test_helpers.py
[the remark this answers is not in the document]
Comment (raulmarindev) [reply openmodels/test_helpers.py:1]: Yep, it makes sense to have it close to our tests. I've created an issue for it: https://github.com/SF-Tec/openmodels/issues/15

**Review by paucablop — APPROVED** [review]
@raulmarindev it looks really good! thanks for the great work! :rocket:

**Pull request merged**